CI: Use fake external target in LVH-based workflows

gentoo-root opened the pull request:
```release-note
Use fake external target in LVH-based workflows.
```

gentoo-root committed:
.github/workflows: Use fake external targets in LVH-based workflows

Recently, we started replacing real external targets (such as
one.one.one.one) in test workflows with fake ones to increase robustness
in case the servers that we don't control go down.

Commit 6982ed921b11 (".github/workflows: Use fake targets in
conformance-kind-proxy-embedded") mentions that the new GitHub actions
kind-external-network and kind-external-targets don't work with
little-vm-helper. The reason is that those actions, as implemented
initially, called into Docker from the GHA context, but Docker runs
inside a VM in LVH-based workflows. The change made here allows those
GitHub actions to SSH into the VM before running Docker commands.

To support Docker volume mounts both on the host and inside the VM, use
relative paths and cd /host in the VM to get into the same directory.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
.github/workflows: Canonicalize an IPv6 address

Canonicalize IP6EXTERNALRANGE to fix this warning:

[=] [cilium-test-2] Test [all-egress-deny-knp] [3/21]
I0526 13:40:09.838874       1 warnings.go:110] "Warning: spec.egress[0].to[1].ipBlock.cidr: IPv6 CIDR value \"fd00:10:64::ffff:00/112\" should be in RFC 5952 canonical format (\"fd00:10:64::ffff:0/112\")"

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
contrib: Support KIND_EXPERIMENTAL_DOCKER_NETWORK in kind.sh

.github/actions/kind-external-network/action.sh passes
KIND_EXPERIMENTAL_DOCKER_NETWORK to kind, but our wrapper script
overwrites it unconditionally and doesn't provide a way to change it.

Make the script check whether KIND_EXPERIMENTAL_DOCKER_NETWORK might be
already set.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
.github/actions: Set bridge name for fake external target

Enabling fake external targets in LVH-based workflows makes this IPv6
NodePort test fail:

north-south-loadbalancing-with-l7-policy-port-range/outside-to-nodeport

tcp_checksum_complete fails in tcp_v6_rcv, and TCP_MIB_CSUMERRORS
increases with every IPv6 SYNACK.

While it's not clear why it passed in other workflows and how setting
the bridge name explicitly affects the checksums, adding this magic line
is sufficient to make this test pass in LVH-based workflows.

The bridge name chosen is not arbitrary, and it matches the config used
by contrib/scripts/kind.sh.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
.github/actions: Change IPv6 subnet for fake external target

check-encryption-leaks.bt treats everything under fd00::/8 as pod
traffic. This CIDR includes the actual external and node-to-node traffic
in configurations with fake external target.

Change the IPv6 subnet in those configurations to match the default used
by contrib/scripts/kind.sh.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
Use cluster DNS to resolve fake external targets

.github/actions/kind-external-targets/action.sh patches the cluster-wide
DNS to be able to resolve the hostnames of the fake external targets in
the cilium TLD.

Some LVH-based workflows use node-local DNS that doesn't know about the
cilium TLD, therefore, it'll fail to resolve the fake external targets
once they are enabled for LVH-based workflows.

Patch the node-local DNS config to forward requests in the cilium TLD to
the cluster-wide DNS server, which is capable of resolving the hostnames
of the fake external targets.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

dylandreimerink reviewed (approved):
Looks good to me. Thank you for the work

qmonnet commented on .github/workflows/tests-ipsec-upgrade.yaml:
```diff
@@ -262,13 +256,41 @@ jobs:
+      - name: Set up job variables for connectivity tests
+        id: vars-conn
+        run: |
+          EXTRA_CLI_FLAGS=(
```
This set of flags is the same in the three workflows where you add it, right? Would it be possible to set them and export them from the `lvh-kind` action, to avoid repeating this chunk, maybe?

gentoo-root replied:
> This set of flags is the same in the three workflows where you add it, right?

Yes, they are exactly the same everywhere.

> Would it be possible to set them and export them from the lvh-kind action, to avoid repeating this chunk, maybe?

Yes, it's definitely possible... I like it slightly less, because having the formatted string as an output of the lvh-kind action is less flexible compared to individual outputs, and because it implies that lvh-kind knows about connectivity tests (high coupling).

But I can implement it for the sake of reduction of repeated lines of code (I can keep the individual outputs too, they will just not be used).

qmonnet replied:
No strong opinion from my side really, just keep them uncoupled then - you have a valid point.

qmonnet commented on .github/actions/kind-external-network/action.sh:
```diff
@@ -42,6 +42,7 @@ lvh_wrapper docker network create \
     --driver bridge \
     -o com.docker.network.bridge.enable_ip_masquerade=true \
     -o com.docker.network.driver.mtu=$MTU \
+    -o com.docker.network.bridge.name=br-$KINDNETWORK \
```
Why is this fix required, is this due to a previous change in the PR, or is it currently broken in `main`?

gentoo-root replied:
So, it's a mystery... I don't know how setting the bridge name fixes the checksums. When the docker network is created by contrib/scripts/kind.sh, [it sets the bridge name](https://github.com/cilium/cilium/blob/d9d85b52d7b63e20dd25ac1c851385b98bce169f/contrib/scripts/kind.sh#L195). This part was missing in the kind-external-network action, but surprisingly it only causes checksum errors if kind runs in LVH. It didn't fail in other (non-LVH) workflows.

So, I'd say the problem existed before my pull request, it just didn't show up outside of LVH for some reason, and I added LVH workflows in this PR.

Cc: @dylandreimerink — maybe you have an idea why setting the bridge name could be important for checksums o_O

dylandreimerink replied:
>  maybe you have an idea why setting the bridge name could be important for checksums o_O

Wut. I have no clue. Does not sound like it should be related :cry:

pchaigno replied:
Even if we can't figure out the root cause, I think this context should be in the commit description.

gentoo-root replied:
> I think this context should be in the commit description.

It's there, but I'll add more details.

qmonnet reviewed (approved):
Two comments below, but none of them blocking.

You have a merge conflict to address, though.

Thanks!

gentoo-root commented on .github/workflows/tests-ipsec-upgrade.yaml:
```diff
@@ -262,13 +256,41 @@ jobs:
+            "\"--external-target-ipv6-capable\""
+          )
+
+          CONNECTIVITY_TEST_DEFAULTS="${EXTRA_CLI_FLAGS[*]}"
```
TODO: Use `"${EXTRA_CLI_FLAGS[@]@Q}"` for robustness in case some of the arguments contain spaces.

pchaigno commented on .github/actions/kind-external-network/action.sh:
```diff
@@ -11,21 +11,21 @@ lvh_wrapper() {
 }
```
On commit *.github/actions: Change IPv6 subnet for fake external target*

> check-encryption-leaks.bt treats everything under fd00::/8 as pod traffic, and it includes the actual external and node-to-node traffic in configurations with fake external target.

Nit: That doesn't parse :-) I think you mean:

> check-encryption-leaks.bt treats everything under fd00::/8 as pod traffic. This CIDR includes the actual external and node-to-node traffic in configurations with fake external target.

gentoo-root replied:
"It" refers to "everything under fd00::/8". I made the change as you suggested, thanks.

pchaigno commented on .github/actions/kind-external-targets/node-local-dns-patch.txt:
```diff
@@ -0,0 +1,11 @@
+    cilium:53 {
```
On commit *Use cluster DNS to resolve fake external targets*.

The commit is missing the Why. Does this change mean DNS resolution didn't work before for external targets?

gentoo-root replied:
> Does this change mean DNS resolution didn't work before for external targets?

Existing workflows that use fake external targets didn't have the "Install node local DNS" step. Cluster-wide DNS is being patched [here](https://github.com/cilium/cilium/blob/7da06b18294fdc47ff34f57d001b41dff8f1f6b2/.github/actions/kind-external-targets/action.sh#L123-L134) to be able to resolve the fake external targets' hostnames.

I'll add the commit description.

pchaigno commented on .github/actions/kind-external-network/action.sh:
```diff
@@ -1,3 +1,15 @@
+#!/usr/bin/env bash
```
On commit *.github/workflows: Use fake external targets in LVH-based workflows*:

This seems to be the main change and I'm not currently able to review it. I think the commit description (and probably the PR description) need a lot more context.

We already had a bunch of things in place for external targets. Why do we need this PR? Is it because LVH couldn't use external targets before? Why? And in this specific commit, how does the commit title relate to the commit description?

It may be easier to explain each change if this big commit was split into multiple changes.

gentoo-root replied:
> I think the commit description (and probably the PR description) need a lot more context.
> Why do we need this PR? Is it because LVH couldn't use external targets before? Why?

I'm sorry for the confusion, you are absolutely right that I need to write down the context that exists in my head. I updated the commit description, now it answers your questions, and refers to Dylan's original commit that also explains that LVH support was kept back for a follow-up.

> And in this specific commit, how does the commit title relate to the commit description?

The original text was just a note on a particular implementation detail, of course it didn't fully explain what was going on here:

```
To support docker volume mounts both on the host and inside the VM, use
relative paths and cd /host in the VM to get into the same directory.
```

pchaigno reviewed (changes requested):
I probably have less background on the recent work for external targets than other reviewers and, maybe because of that, I'm currently unable to review the main changes here.

I think this pull request and its commits need more context, to help me review and future debuggers understand. I've left comments to try and state what is unclear to me, but in general I'd expect the Why to be stated for each commit and, if the changes don't trivially relate to the commit's title/description, additional explanation for what's non-obvious (cf. Quentin's comment for an example).

gentoo-root committed:
.github/workflows: Safe quoting for EXTRA_CLI_FLAGS

EXTRA_CLI_FLAGS is formed as an array, each element corresponding to a
command line argument. However, if an argument contains spaces, we have
to quote the element of the array manually. Alleviate this burden by
applying shell quoting with @Q when expanding the array into a string.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

pchaigno reviewed (approved):
Thanks a lot Maxim! The new commit descriptions help a lot for review and it all looks good to me :)
